README.md
# Functional-programming based Gross to net salary calculator API

An API written in node.js using functional techniques & Express that calculates your net take-home pay!

## Configuration
The config is currently hard coded in taxRules.js. 🙃

I am trying to think of a clever way to make this deployable as config. 👍

## Running

This project uses NVM to allow you to keep things consistent, I'm using node 10 because this is a personal project trying the new features is a worthwhile trade off for stability. 🤷‍

✅ Select the correct node version using
```
$ nvm use
```
💪 Or, if you don't have the specified version install and use with
```
$ nvm install 10.2.0

$ nvm use
```
🔧 Install
```
$ yarn
```
🎮 Then start
```
$ yarn start
```

It's as easy as that! 😎


## Contributing

Read my [contributing guide](/CONTRIBUTING.md) to learn how to get started.

## Changelog

My [changelog](/CHANGELOG.md) is generated when a release is performed, using standard version for automated docs goodness. 💅


## Technical design considerations

The model currently is based on the real UK tax rules as per [gov.uk](https://www.gov.uk/browse/tax/income-tax) as interpreted by myself; I am not an accountant! 🤓

# Disclaimer
## I am not an accountant!
### This software is presented with no liability or assumption of any sort of accounting expertise on its behalf
I have spent the bulk of my time thinking about how to make this project functional, and node 10 rather than how tax should work, so please don't base any real estimations off my assumptions. 😱


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.0"></a>
# 1.0.0 (2018-06-03)


### Bug Fixes

* **lint:** added missing lint job ([025cc98](https://github.com/AlexanderCannon/tax-calculator/commit/025cc98))
* **linting job:** fixed linting script ([301c89a](https://github.com/AlexanderCannon/tax-calculator/commit/301c89a))
* **national insurance:** improved national insurance accuracy ([35dab95](https://github.com/AlexanderCannon/tax-calculator/commit/35dab95))


### Features

* **fastify:** upgraded express to fastify ([3a0ece9](https://github.com/AlexanderCannon/tax-calculator/commit/3a0ece9))
* **national insurance:** added accurate national insurance calculation ([aa7c727](https://github.com/AlexanderCannon/tax-calculator/commit/aa7c727))
* **net income:** added net income to return value ([98b1c30](https://github.com/AlexanderCannon/tax-calculator/commit/98b1c30))
* **pennies!:** converted the service to use pennies rather than pounds ([514e24a](https://github.com/AlexanderCannon/tax-calculator/commit/514e24a))
* **ramda:** used ramda to make code safer and more elegant ([5690fc4](https://github.com/AlexanderCannon/tax-calculator/commit/5690fc4))
* **student loan:** added student loan calculation ([f612af1](https://github.com/AlexanderCannon/tax-calculator/commit/f612af1))


### BREAKING CHANGES

* **national insurance:** API now takes DoB instead of age int

Before:
```
{
  "grossIncome": "18500",
  "age": "28",
  "pensionContributions": "0.0",
  "studentLoanPlan": "plan1"
}
```

Now:
```
{
  "grossIncome": "18500",
  "dateOfBirth": "1989/7/17",
  "pensionContributions": "0.0",
  "studentLoanPlan": "plan1"
}
```



<a name="0.2.0"></a>
# [0.2.0](https://github.com/AlexanderCannon/tax-calculator/compare/v0.1.1...v0.2.0) (2018-05-28)


### Features

* **pennies!:** converted the service to use pennies rather than pounds ([514e24a](https://github.com/AlexanderCannon/tax-calculator/commit/514e24a))
* **ramda:** used ramda to make code safer and more elegant ([5690fc4](https://github.com/AlexanderCannon/tax-calculator/commit/5690fc4))



<a name="0.1.1"></a>
## 0.1.1 (2018-05-27)


### Bug Fixes

* **lint:** added missing lint job ([025cc98](https://github.com/AlexanderCannon/tax-calculator/commit/025cc98))


src/utils/tax.test.js
const {
  makePennies,
  fromPennies,
  getOwedForBand,
  uncappedBand,
  getBandCap,
  getBandStart,
  bandStartLowerThanAllowance,
  makeMonthly,
  makeWeekly,
  makeDaily,
} = require('./tax');

const taxLowerBand = {
  start: 1185900,
  rate: 0.20,
  end: 4635000,
};
const taxHigherBand = {
  start: 4635000,
  end: 15000000,
  rate: 0.40,
};
const taxEndBand = {
  start: 15000000,
  end: -1,
  rate: 0.45,
};
const hidden = {
  one: 1,
  // const pensionAge = 65;
  // const niFree = {
  //   start: 0.00,
  //   end: 162.00,
  //   rate: 0.00,
  // };
  // const niLower = {
  //   start: 162.00,
  //   end: 892.00,
  //   rate: 0.12,
  // };
  // const niHigher = {
  //   start: 892.00,
  //   end: -1,
  //   rate: 0.02,
  // };

  // const studentLoanPlan1 = {
  //   threshold: 18330.00,
  //   rate: 0.09,
  // };
  // const studentLoanPlan = {
  //   threshold: 25000.00,
  //   rate: 0.09,
  // };
};
hidden.one = 2;

const salaryBelowMinIncome = 1185000;
const salaryEqualMinIncome = 1185900;
const salaryAboveMinIncome = 1186000;
const salaryStartBracketMid = 2400000;
const salaryStartBracketEnd = 4635000;
const salaryMidBracketStart = 4635100;
// const salaryMidBracketEnd = 150000;
// const salaryEndBracketStart = 150001;
const salaryEndBracketHigh = 25000000;

describe('makePennies', () => {
  it('should return 100 times the value given', () => {
    expect(makePennies(100)).toEqual(10000);
    expect(makePennies(1)).toEqual(100);
    expect(makePennies(25)).toEqual(2500);
    expect(makePennies(0.01)).toEqual(1);
    expect(makePennies(0.10)).toEqual(10);
    expect(makePennies(0.03)).toEqual(3);
  });
});

describe('fromPennies', () => {
  it('should divide the value by 100', () => {
    expect(fromPennies(10000)).toEqual(100.00);
    expect(fromPennies(100)).toEqual(1.00);
    expect(fromPennies(2500)).toEqual(25.00);
    expect(fromPennies(1)).toEqual(0.01);
    expect(fromPennies(10)).toEqual(0.10);
    expect(fromPennies(3)).toEqual(0.03);
  });
});

describe('getBandStart', () => {
  it('should return 0 if salary is lower than the band min', () =>
    expect(getBandStart(salaryBelowMinIncome)(taxLowerBand.start)).toEqual(0));
  it('should return 0 if salary equals the band min', () =>
    expect(getBandStart(salaryEqualMinIncome)(taxLowerBand.start)).toEqual(0));
  it('should not change the band if salary is in than the band min', () =>
    expect(getBandStart(salaryAboveMinIncome)(taxLowerBand.start))
      .toEqual(taxLowerBand.start));
});

describe('getBandCap', () => {
  it('should return the band cap if the band cap is less than the salary', () =>
    expect(getBandCap(salaryMidBracketStart)(taxLowerBand.end))
      .toEqual(taxLowerBand.end));
  it('should return the salary if the band cap is less than the salary', () =>
    expect(getBandCap(salaryStartBracketMid)(taxLowerBand.end))
      .toEqual(salaryStartBracketMid));
  it('should return the band cap if the band cap equals the salary', () =>
    expect(getBandCap(salaryStartBracketEnd)(taxLowerBand.end))
      .toEqual(taxLowerBand.end));
  it('should return the salary if the band cap is -1', () =>
    expect(getBandCap(salaryStartBracketEnd)(taxEndBand.end))
      .toEqual(salaryStartBracketEnd));
});

describe('getOwedForBand', () => {
  it('should return accurately calculated tax for a given band', () =>
    expect(getOwedForBand(taxLowerBand.start, taxLowerBand.end, taxLowerBand.rate))
      .toEqual(689820));
  it('should be accurate for the lower rate with no allowance', () =>
    expect(getOwedForBand(0, taxLowerBand.end, taxLowerBand.rate))
      .toEqual(927000));
  it('should be accurate for the higher rate of tax', () =>
    expect(getOwedForBand(taxHigherBand.start, taxHigherBand.end, taxHigherBand.rate))
      .toEqual(4146000));
  it('should be accurate for the end rate of tax', () =>
    expect(getOwedForBand(taxEndBand.start, salaryEndBracketHigh, taxEndBand.rate))
      .toEqual(4500000));
});

describe('bandStartLowerThanAllowance', () => {
  it('should return the allowance if it is higher than the band start', () =>
    expect(bandStartLowerThanAllowance(salaryAboveMinIncome)(taxLowerBand.start))
      .toEqual(salaryAboveMinIncome));
  it('should return the allowance if the band start is 0', () =>
    expect(bandStartLowerThanAllowance(salaryAboveMinIncome)(0))
      .toEqual(salaryAboveMinIncome));
  it('should return the allowance if the band start is negative', () =>
    expect(bandStartLowerThanAllowance(salaryAboveMinIncome)(-1))
      .toEqual(salaryAboveMinIncome));
  it('should return the band start allowance is smaller', () =>
    expect(bandStartLowerThanAllowance(salaryBelowMinIncome)(taxHigherBand.start))
      .toEqual(taxHigherBand.start));
});

describe('makeMonthly', () => {
  it('should return 1/12th of the value', () => {
    expect(makeMonthly(12)).toEqual(1);
    expect(makeMonthly(3400)).toEqual(283.3333333333333);
    expect(makeMonthly(24000)).toEqual(2000);
  });
});

describe('makeWeekly', () => {
  it('should return 1/12th of the value', () => {
    expect(makeWeekly(52)).toEqual(1);
    expect(makeWeekly(3400)).toEqual(65.38461538461539);
    expect(makeWeekly(24000)).toEqual(461.53846153846155);
  });
});

describe('makeDaily', () => {
  it('should return 1/12th of the value', () => {
    expect(makeDaily(365)).toEqual(1);
    expect(makeDaily(3400)).toEqual(9.315068493150685);
    expect(makeDaily(24000)).toEqual(65.75342465753425);
  });
});
describe('uncappedBand', () => {
  it('should return true if band cap is less than 0', () =>
    expect(uncappedBand(-1)).toEqual(true));
  it('should return false if band cap is 0', () =>
    expect(uncappedBand(0)).toEqual(false));
  it('should return false if band cap is positive', () => {
    expect(uncappedBand(1)).toEqual(false);
    expect(uncappedBand(salaryAboveMinIncome)).toEqual(false);
  });
});


src/utils/index.test.js
const {
  nearestWholePenny,
  divideRightToLeft,
  curriedDivideRightToLeft,
  pickSmaller,
  curriedPickSmaller,
  pickLarger,
  curriedPickLarger,
  aLessThanB,
  curriedALessThanB,
  getAge,
  addThree,
  curriedAddThree,
} = require('./');

describe('nearestWholePenny', () => {
  it('should return a rounded number', () => {
    expect(nearestWholePenny(1.000000098)).toEqual(1);
    expect(nearestWholePenny(1.88888889)).toEqual(1.89);
  });
  it('should not round a whole number', () => {
    expect(nearestWholePenny(1)).toEqual(1);
    expect(nearestWholePenny(2)).toEqual(2);
  });
});

describe('divideRightToLeft', () => {
  it('should divide the params in the opposite order they are passed', () =>
    expect(divideRightToLeft(2, 1)).toEqual(0.5));
});

describe('curriedDivideRightToLeft', () => {
  it('should divide the params in the opposite order they are passed', () =>
    expect(curriedDivideRightToLeft(2)(1)).toEqual(0.5));
});

describe('pickSmaller', () => {
  it('pick the first param if it is smaller', () =>
    expect(pickSmaller(1, 2)).toEqual(1));
  it('pick the second param if it is smaller', () =>
    expect(pickSmaller(2, 1)).toEqual(1));
});

describe('curriedPickSmaller', () => {
  it('pick the first param if it is smaller', () =>
    expect(curriedPickSmaller(1)(2)).toEqual(1));
  it('pick the second param if it is smaller', () =>
    expect(curriedPickSmaller(2)(1)).toEqual(1));
});
describe('pickLarger', () => {
  it('pick the first param if it is larger', () =>
    expect(pickLarger(2, 1)).toEqual(2));
  it('pick the second param if it is larger', () =>
    expect(pickLarger(1, 2)).toEqual(2));
});

describe('curriedPickLarger', () => {
  it('pick the first param if it is larger', () =>
    expect(curriedPickLarger(2)(1)).toEqual(2));
  it('pick the second param if it is larger', () =>
    expect(curriedPickLarger(1)(2)).toEqual(2));
});

describe('aLessThanB', () => {
  it('should return true if a is less than b', () =>
    expect(aLessThanB(1, 2)).toBe(true));
  it('should return true if a is less than b', () =>
    expect(aLessThanB(2, 1)).toBe(false));
});

describe('curriedALessThanB', () => {
  it('should return true if a is less than b', () =>
    expect(curriedALessThanB(1)(2)).toBe(true));
  it('should return true if a is less than b', () =>
    expect(curriedALessThanB(2)(1)).toBe(false));
});

describe('getAge', () => {
  it('should return the correct age', () => {
    expect(getAge('1989/07/17', '2018/6/3')).toEqual(28);
    expect(getAge('2000/07/17', '2018/7/17')).toEqual(18);
    expect(getAge('2000/07/17', '2018/7/18')).toEqual(18);
    expect(getAge('2000/07/17', '2018/7/16')).toEqual(17);
  });
  it('should not return less than the correct age at the time of writing', () => {
    expect(getAge('1989/07/17')).toBeGreaterThanOrEqual(28);
    expect(getAge('2000/07/17')).toBeGreaterThanOrEqual(17);
  });
});

describe('addThree', () => {
  it('should add three numbers', () => {
    expect(addThree(1, 1, 1)).toEqual(3);
    expect(addThree(1, 2, 3)).toEqual(6);
  });
});
describe('curriedAddThree', () => {
  it('should add three numbers when passed one param then two', () => {
    expect(curriedAddThree(1)(1, 1)).toEqual(3);
    expect(curriedAddThree(1)(2, 3)).toEqual(6);
  });
  it('should add three numbers when passed two param then one', () => {
    expect(curriedAddThree(1, 1)(1)).toEqual(3);
    expect(curriedAddThree(1, 2)(3)).toEqual(6);
  });
  it('should add three numbers when called three times', () => {
    expect(curriedAddThree(1)(1)(1)).toEqual(3);
    expect(curriedAddThree(1)(2)(3)).toEqual(6);
  });
  it('should be a function if it is passed two the params', () => {
    expect(typeof curriedAddThree(1)(1)).toBe('function');
    expect(typeof curriedAddThree(1, 2)).toBe('function');
  });
  it('should be a function if it is passed one param', () => {
    expect(typeof curriedAddThree(1)).toBe('function');
  });
});


src/utils/tax.js
const {
  curry,
  multiply,
  subtract,
  gt,
} = require('ramda');
const { curriedDivideRightToLeft, pickSmaller, aLessThanB } = require('./');

const makePennies = multiply(100);
const fromPennies = curriedDivideRightToLeft(100);

const getOwedForBand = (start, end, rate) => (
  aLessThanB(start, end)
    ? multiply(subtract(end, start), rate)
    : 0
);

const curriedGetOwedForBand = curry(getOwedForBand);

const uncappedBand = bandEnd => bandEnd === -1;

const getBandEnd = (salary, bandEnd) => (
  uncappedBand(bandEnd)
    ? salary
    : pickSmaller(salary, bandEnd)
);

const curriedGetBandEnd = curry(getBandEnd);

const getBandStart = (salary, bandStart) => (
  gt(salary, bandStart)
    ? bandStart
    : 0
);

const curriedGetBandStart = curry(getBandStart);

const bandStartLowerThanAllowance = (allowance, bandStart) => (
  gt(allowance, bandStart)
    ? allowance
    : bandStart
);

const curriedBandStartLowerThanAllowance = curry(bandStartLowerThanAllowance);

const makeMonthly = curriedDivideRightToLeft(12);
const makeWeekly = curriedDivideRightToLeft(52);
const makeDaily = curriedDivideRightToLeft(365);

const fromMonthly = multiply(12);
const fromWeekly = multiply(52);
const fromDaily = multiply(365);

module.exports = {
  makePennies,
  fromPennies,
  getOwedForBand,
  curriedGetOwedForBand,
  uncappedBand,
  getBandEnd,
  curriedGetBandEnd,
  getBandStart,
  curriedGetBandStart,
  bandStartLowerThanAllowance,
  curriedBandStartLowerThanAllowance,
  makeMonthly,
  makeWeekly,
  makeDaily,
  fromMonthly,
  fromWeekly,
  fromDaily,
};


src/utils/personalAllowance.js
const { curry } = require('ramda');

const { nearestWholePenny } = require('./');

const getAgeRelatedContributions = (allowance, user) => {
  if (user.age >= 75) {
    return allowance.basic - allowance.age75over;
  }
  if (user.age >= 65) {
    return allowance.basic - allowance.age6574;
  }
  return 0;
};

const getAgeRelatedTaperDeductions = (
  allowance,
  user,
  grossIncome,
) => {
  const incomeMinusTaperThreshold =
    grossIncome - allowance.thresholds.taper;
  if (incomeMinusTaperThreshold < 0) return 0;

  const halfIncomeMinusTaperThreshold = incomeMinusTaperThreshold / 2;
  const ageRelatedContributions = getAgeRelatedContributions(allowance, user);
  if (halfIncomeMinusTaperThreshold > ageRelatedContributions) return ageRelatedContributions;

  return halfIncomeMinusTaperThreshold;
};

const getAllowanceAfterAgeAdjust = (
  allowance,
  user,
  grossIncome,
) => {
  const ageAllowance =
    allowance.basic + getAgeRelatedContributions(allowance, user);
  return ageAllowance - getAgeRelatedTaperDeductions(allowance, user, grossIncome);
};


const getTaperDeductions = (
  allowance,
  user,
  grossIncome,
) => {
  const incomeMinusPensionContributions =
    grossIncome - user.pensionContributions;
  const incomeMinusPensionMinusTaperThreshold =
    incomeMinusPensionContributions - allowance.thresholds.taper;
  if (incomeMinusPensionMinusTaperThreshold < 0) return 0;

  const halfIncomeMinusPensionMinusTaperThreshold =
    incomeMinusPensionMinusTaperThreshold / 2;
  const allowanceAfterAgeAdjust = getAllowanceAfterAgeAdjust(allowance, user, grossIncome);
  if (halfIncomeMinusPensionMinusTaperThreshold > allowanceAfterAgeAdjust) {
    return allowanceAfterAgeAdjust;
  }
  return halfIncomeMinusPensionMinusTaperThreshold;
};

const getPersonalAllowance = (
  allowance,
  user,
  grossIncome,
) =>
  nearestWholePenny(getAllowanceAfterAgeAdjust(allowance, user, grossIncome)
    - getTaperDeductions(allowance, user, grossIncome));

const getBlindAllowance = (allowance, user) => {
  if (!user.blind) return 0;

  return allowance.blind;
};

const getTotalAllowances = (
  grossIncome,
  allowance,
  user,
) => getPersonalAllowance(allowance, user, grossIncome) + getBlindAllowance(allowance, user);

const curriedTotalAllowances = curry(getTotalAllowances);

module.exports = {
  getAgeRelatedContributions,
  getAgeRelatedTaperDeductions,
  getAllowanceAfterAgeAdjust,
  getTaperDeductions,
  getPersonalAllowance,
  getBlindAllowance,
  getTotalAllowances,
  curriedTotalAllowances,
};


src/taxRules.js
module.exports = [
  {
    year: '2018/19',
    pensionAge: 65,
    allowances: {
      basic: 1185900,
      age6574: 1185900,
      age75over: 1185900,
      blind: 232000,
      thresholds: {
        age: 2770000,
        taper: 10000000,
      },
    },
    incomeTax: [
      {
        start: 0,
        rate: 0.20,
        end: 4635000,
      },
      {
        start: 4635000,
        end: 15000000,
        rate: 0.40,
      },
      {
        start: 15000000,
        end: -1,
        rate: 0.45,
      },
    ],
    nationalInsurance: [
      {
        start: 0,
        end: 16200,
        rate: 0,
      },
      {
        start: 16200,
        end: 89200,
        rate: 0.12,
      },
      {
        start: 89200,
        end: -1,
        rate: 0.02,
      },
    ],
    studentLoan: {
      plan1: {
        threshold: 1833000,
        rate: 0.09,
      },
      plan2: {
        threshold: 2500000,
        rate: 0.09,
      },
    },
  },
  {
    year: '2017/18',
    pensionAge: 65,
    allowances: {
      basic: 1150000,
      age6574: 1150000,
      age75over: 1150000,
      blind: 239000,
      thresholds: {
        age: 2770000,
        taper: 10000000,
      },
    },
    incomeTax: [
      {
        start: 0,
        end: 3350000,
        rate: 0.20,
      },
      {
        start: 3350000,
        end: 15000000,
        rate: 0.40,
      },
      {
        start: 15000000,
        end: -1,
        rate: 0.45,
      },
    ],
    nationalInsurance: [
      {
        start: 0,
        end: 15700,
        rate: 0,
      },
      {
        start: 15700,
        end: 86600,
        rate: 0.12,
      },
      {
        start: 86600,
        end: -1,
        rate: 0.02,
      },
    ],
    studentLoan: {
      plan1: {
        threshold: 1777500,
        rate: 0.09,
      },
      plan2: {
        threshold: 2100000,
        rate: 0.09,
      },
    },
  },
];
